warc_metadata_sidecar.py

```diff
@@ -0,0 +1,279 @@
+#!/usr/bin/python
+import argparse
+import io
+import logging
+import os
+import pkg_resources
+import re
+import regex
+import socket
+import time
+from datetime import timedelta
+
+import chardet
+import pycld2 as cld2
+from fido.fido import Fido
+from magic.compat import detect_from_content
+from warcio.archiveiterator import ArchiveIterator
+from warcio.warcwriter import WARCWriter
+
+
+MIME_TITLE = 'Identified-Payload-Type:'
+METHOD_TITLE = 'Payload-Type-Retrieval-Method:'
+RETRIEVED_TITLE = 'Payload-Type-Retrieved-By:'
+PUID_TITLE = 'Preservation-Identifier:'
+CHARSET_TITLE = 'Charset-Detected:'
+LANGUAGE_TITLE = 'Languages-cld2:'
+
+BAD_CHARS = regex.compile(r'\p{Cc}|\p{Cs}|\p{Cn}')
+
+
+class ExtendFido(Fido):
+    """A class that extends Fido to override some methods."""
+    def blocking_read(self, file, bytes_to_read):
+        """Read all bytes into buffer and return.
+
+        Remedies a known issue of the method hanging when identify_stream is called.
+        Modifies the if statement to break out when the end of the stream is reached.
+        https://github.com/openpreserve/fido/blob/093cf9c8c968c710d3d6dfbbcc6e067cd9e27ef3/fido/fido.py#L518
+        """
+        bytes_read = 0
+        buffer = b''
+        while bytes_read < bytes_to_read:
+            readbuffer = file.read(bytes_to_read - bytes_read)
+            buffer += readbuffer
+            bytes_read = len(buffer)
+            if not readbuffer:  # Also works with if readbuffer == b'':
+                break
+        return buffer
+
+    def print_matches(self, fullname, matches, matchtype=''):
+        """Override print_matches to store a mimetype and puid to the fido object."""
+        # https://github.com/openpreserve/fido/blob/master/fido/fido.py#L272
+        if not matches:
+            self.mime = None
+            self.puid = None
+            self.matchtype = matchtype
+            self.format_tool = 'Fido'
+            return
+        match_format, _ = matches[0]
+        mime = match_format.find('mime')
+        puid = None
+        self.mime = mime.text if mime is not None else None
+        # We only want puid from the content.
+        if matchtype != 'extension':
+            puid = self.get_puid(match_format)
+        self.puid = puid if puid is not None else None
+        self.matchtype = matchtype
+        self.format_tool = 'Fido'
```

Comment:
Because of the several changes we might move forward with, I wondered what you thought about me getting rid of the `print_matches` method and doing all the logic in `identify_stream` (that I am overriding anyway). It would look like this:

```
def identify_stream(self, stream):
     """Override identify_stream to get the matches, mime type, and puid"""
     bofbuffer, eofbuffer, bytes_read = self.get_buffers(stream)
     self.current_filesize = bytes_read
     matches = self.match_formats(bofbuffer, eofbuffer)
     self.puid = None
     self.mime = None
     if matches:
         match_format, _ = matches[0]
         mime = match_format.find('mime')
         self.mime = mime.text if mime is not None else None
         self.puid = self.get_puid(match_format)

Reply:
I think it makes sense to do that.